infra/app/build-repository/main.tf

```diff
@@ -0,0 +1,51 @@
+locals {
```

Comment:
Creating a root module for single resource seems counter-intuitive to me. I believe this module should be called in `app/envs/prod` because, despite being needed by multiple environments it is a production resource.

Reply:
we discussed this in the meeting with @kyeah. I don't think the module architecture should be based on number of resources, but rather it should be based on what is conceptually easier to understand. if this is called in `app/envs/prod`, then

1. `prod`, `stage/`, and `dev` will no longer be symmetrical and `prod` will contain an extra resource that isn't obvious to someone who doesn't know that
2. it means developers have to create `prod` _before_ they can create `dev` or `stage`, which is not intuitive or what needs to happen in practice. in most projects people spin up their `dev` environment first, then their `stage` environment, then `prod` last.

having environments depend on cross-env resources like the build repository, and having all resources depend on account resources like tf backends, is a natural linear dependency relationship that i think is simple to understand and i think that simplicity is well worth needing to do one extra step of having to deploy one extra root module.

Reply:
I recall the conversation and I don’t think the architecture should be based on number of modules either. However, as it stands, we have dangling resource that if I came in as infra engineer new to the project, I would have to investigate why this resource has been separated as it has been. What about applications that have more than one image? Further, does this now mean we create a new root for every cross-environment resource? 

In regards to bullets:

1.	It does not matter if they are symmetrical and there’s a good chance they wont remain symmetrical. There may be resources needed in higher environments that are not needed in lower or vice versa. The prod account still has an extra resource it is just referenced somewhere else. 
2.	We’re creating this resource only in the prod account. We’re just obfuscating where the resource is created. Honestly this bullet point pushes me further towards the argument that this should be managed under `accounts/account/maint.tf > module “app1_ecr”, module “app2_ecr”` and so forth.

Thinking about this further this architecture creates issues for other resources. The directory structure now separates everything by app. However, when it comes to resources like ECS, you will have a single cluster per environment with multiple applications running in it. Now, we have nowhere to manage that, unless we stick it in account/accounts, which I would not do because ECS is generally managed under an environment. I think we are trying to make this template do too many things. Right now, we have a (mono-multi)repo-(single-multi)account-multi-environment-(single-multi)-application architecture and I think the patterns are beginning to conflict. I wonder if we should scope this down and tackle just a simple single application infrastructure and then expand up from there.

Reply:
> I wonder if we should scope this down and tackle just a simple single application infrastructure and then expand up from there.

+1 Definitely. This repo is currently only intended to be for a single-application repo.

> However, when it comes to resources like ECS, you will have a single cluster per environment with multiple applications running in it. 

I'll admit I don't know much about ECS cluster best practices. Any issues with having a separate cluster for each application + environment cluster (for example, what are your thoughts on having it called from the app/env-template module)? Is it costly to have multiple clusters / is it important to minimize the number of ECS clusters created?

> What about applications that have more than one image?

I was imagining that applications that have more than one image repository would put that image repository in the `app/build-repository` root module along with the other one. The build-repository contains the builds for the application.

> We’re just obfuscating where the resource is created.

I'm a little confused by this point because I think it's the opposite. As a developer, if I see the following folder structure, it's not obvious to me where the ECR resource gets created. ECR is an application resource specific to storing an application's container images, so I'd expect it to live somewhere in the app folder, but it also needs to be shared across environments so I wouldn't know where it lives.
```
infra/
  accounts/
    prod
    dev
  app
    env-template/
    envs/
      prod
      dev
```

on the other hand, if i see a folder called `app/build-repository` I could infer that that's where ECR is created. Or for a static website, that's where the S3 bucket storing static site builds would live.

One other factor is that I think it's important to separate out sensitive resources that set up foundational trust items like OIDC, so I really don't want to mix ECR and ECS into the same root module as OIDC. I remember you saying that you're used to having AdministratorAccess but I don't think that's something we can typically expect, so I think it's important to separate out application resources (ECR, ECS) from infrastructure resources (OIDC, IAM). I believe there's a high chance on many projects that developers won't able to `terraform apply` the account root modules themselves and that they'd have to ask some other central IT team to provision those resources for them. So I'm hoping to keep those more sensitive account resources separate from the application level resources that I believe require less trust / permissions to create.

I have an idea – what do you think about doing some sort of user test? At the end of the day, our goal is to make the architecture easy to understand and use by engineers, so I'm thinking it could be valuable to find some way for engineers to actually try out the various options we're proposing and see which one results in a better developer experience.

Reply:
Another idea / possibility I'd ask is if we could consider whether this discussion is blocking, or whether we can revisit later. I remember @kyeah saying that this isn't a huge decision, so if the two of you are comfortable moving forward with the current setup, I'd really like to get to an MVP where we have ECS and the database set up and working with template-application-flask, and I'm happy to revisit this later.

Reply:
1.  If this repo was intended for single application I am not sure why we redesigned the directory structure to add app/ as the top level. 
2. ECS: Best practice is a cluster with multiple apps running inside it, it makes it easier to manage and while the cost of the cluster itself isn't more expensive by have multiple single app clusters peripheral charges could potentially add up. Generally the more infrastructure you create the more complicated it becomes and the more difficult it becomes to manage, it makes harder for patching, upgrades,  figuring what broke when something does break, etc etc.
3. Sorry I meant abstracting where the infrastructure is managed. It's not clear that it's managed in the production account and it seems like we are just abstracting for the sake of it. Permissions aside I wouldn't want to stick it under env/prod either. I might create a shared/ root or global/ root that is clearly defined as being linked to the production account. I would have to think about it more.
4. ECR and ECS are infrastructure, the applications image is stored in the ECR infra and containers run in the ECS infra. When I am looking at infra I want to know WHERE my infrastructure is running, especially when it breaks.
5. Dividing infrastructure potential access levels doesn't make sense to me, what is the benefit of requiring multiple people to manage and deploy infra? Further to setup ECR (and more) policies you're going to need IAM permissions. I have not necessarily had the specific AWS managed AdministratorAccess, but I have always had the appropriate permissions required to deploy and manage infrastructure. 
6. In regards to the user test, who is the users are you saying that this template is meant to eliminate the need for an infrastructure engineer? 
7. No I don't think this is a blocking issue and not like we cannot create more repos or refine the design later.

Reply:
> 1. If this repo was intended for single application I am not sure why we redesigned the directory structure to add app/ as the top level.

Are you referring to `app` at the project root, or `infra/app`? `app` at the project root is needed in order to separate app code from other things like docs and infra, and also to be able to have a relatively clean separation of what's in template-application-(nextjs/flask) vs what's in template-infra. `infra/app` is to have a parallel structure with the top level `app`, and also to allow project teams to add multiple apps on their own if they want (they template won't by default attempt to solve multiple-app, but we also don't want to prevent it unnecessarily)

> 2. ECS
Thanks for explaining that. I think that makes sense. Question - Are a cluster's resources shared? In other words, if one application was acting up and taking up a lot of CPU/RAM/doing lots of disk and network IO, could that potentially impact other applications running on the same cluster? I'm kind of curious what the concept of a cluster even means within AWS, vs just something like Lambda that just runs.

> 3/4. Sorry I meant abstracting where the infrastructure is managed.
Oh I see, you mean what environment the infrastructure is in. Sorry for the confusion. It sounds like "where the infrastructure is managed" can have two meanings – "where it's managed in the code", and "what environment/account it's managed in", and I misunderstood which one you were talking about. I see your point that it's easier to know that it's in the prod environment if it's in accounts/prod.

> 5. Dividing infrastructure potential access levels doesn't make sense to me, what is the benefit of requiring multiple people to manage and deploy infra?
Fair points, let's see what it looks like in practice with NJ UI and DOL ARPA so we can ground the discussion with some examples, as I may be overindexing on the experience from PFML. FWIW I'm not arguing that we should be requiring multiple people to manage and deploy infra, but sometimes those requirements get imposed on us regardless. On every gov project I've been at, the agency had wanted a "separation of concerns" where different classes of engineers have different permissions (one of the most common examples I've seen is the policy that "people who can check code in shouldn't be allowed to deploy code", which is something we spend a lot of energy to correct, so I want us to be as flexible as we can be so that we can pick our battles).

> 6. In regards to the user test, who is the users? Are we trying to get around the need for an infrastructure engineer?

Ideally we'd test with a few different user types. infra engineers; developers; tech leads;

> 7. Are you wanting to store the state in the prod account but deploy the ECR in the dev account? ... If we put ECR in accounts/account-where-environment-lives/main.tf we know beyond a shadow of doubt that is where that infrastructure resides.
Yeah it's a good point that it's much less clear about what account the build-repository resources live in. You have to inspect the terraform backend and just notice that the tfstate bucket is the prod bucket and then infer that it's therefore living in the prod account. If you don't think we should put it in `env/prod` either, I'm open to the idea you're proposing about some sort of naming and organization that is clear about things being in production. I'm also open to revisiting whether we have top level `infra/envs` instead of putting `envs` under `app`

> 8

Thanks for the discussion, let's come back to this and keep refining this design until we are really happy with it